src/model/ActionSet.cs

```diff
@@ -8,38 +8,86 @@ namespace tcg
 {
   class ActionSet
   {
+    private static Player attacker;
+    private static Player target;
+
+    private static Card attackerCard;
+    private static Card targetCard;
+
     public static Dictionary<ActionType, Func<GameState, GameState>> actions = new Dictionary<ActionType, Func<GameState, GameState>>
         {
             {ActionType.Attack, Attack },
+            {ActionType.HealSelf, HealSelf},
+            {ActionType.Die, Die},
+            {ActionType.Print, Print}
         };
 
     private static GameState Attack(GameState state)
     {
-      // please, do not use "var"
-      var attacker = state.CurrentPlayer;
-      var target = state.Players[0].Id != attacker.Id ? state.Players[0] : state.Players[1];
+      GetPlayersAndCards(state);
 
-      var attackerCard = attacker.ActiveCards[state.Attacker];
-      var targetCard = target.ActiveCards[state.Target];
-
-      var newAttackerHP = attackerCard.HP - targetCard.Attack;
-      var newTargetHP = targetCard.HP - attackerCard.Attack;
+      if (attackerCard.UseAction != null)
+      {
+        ActionType type = attackerCard.UseAction.Type;
+        state = actions[type](state);
+      }
 
       attackerCard.HP -= targetCard.Attack;
       targetCard.HP -= attackerCard.Attack;
 
-      // i suppose we can use Class instead of Struct for Card, Player etc
-      // to access them by link
-      // anyway, we are not goind to do completely immutable system
-      // so we can use mutations where it's the easiest strategy
-      attacker.ActiveCards[0] = new Card(attackerCard.ManaCost, newAttackerHP, attackerCard.Attack);
-      target.ActiveCards[0] = new Card(targetCard.ManaCost, newTargetHP, targetCard.Attack);
-
-      // we are returning "state" here to have an ability to chain actions
-      // GameState freshState = Action1(Action2(Action3(state)))
-      // maybe we need to build "pipe" function to make it more clear
-      // GameState freshState = Pipe(state, Action1, Action2, Action3)
+
+      return state;
+    }
+
+    private static GameState HealSelf(GameState state)
+    {
+      GetPlayersAndCards(state);
+
+      attackerCard.HP += attackerCard.UseAction.Power;
+
+      return state;
+    }
+
+    private static GameState Die(GameState state)
+    {
+      GetPlayersAndCards(state);
+      CheckAndPerformDeath(state);
       return state;
     }
+
+    private static void CheckAndPerformDeath(GameState state)
+    {
+      foreach (Player player in state.Players)
+      {
+        for (int i = 0; i < player.ActiveCards.Count; i++)
+        {
+          Card card = player.ActiveCards[i];
+          if (card.HP <= 0)
+          {
+            if (card.DieAction != null)
+            {
+              ActionType type = card.DieAction.Type;
+              state = actions[type](state);
+            }
+            player.ActiveCards.RemoveAt(i);
```

Comment:
What will happen if `DieAction` changes the order of cards in `ActiveCards` :thinking: